recipe-client-addon/bootstrap.js

```diff
@@ -4,95 +4,27 @@
+  "resource://shield-recipe-client/lib/ShieldRecipeClient.jsm");
 
-const REASONS = {
-  APP_STARTUP: 1,      // The application is starting up.
-  APP_SHUTDOWN: 2,     // The application is shutting down.
-  ADDON_ENABLE: 3,     // The add-on is being enabled.
-  ADDON_DISABLE: 4,    // The add-on is being disabled. (Also sent during uninstallation)
-  ADDON_INSTALL: 5,    // The add-on is being installed.
-  ADDON_UNINSTALL: 6,  // The add-on is being uninstalled.
-  ADDON_UPGRADE: 7,    // The add-on is being upgraded.
-  ADDON_DOWNGRADE: 8,  // The add-on is being downgraded.
-};
-
-const PREF_BRANCH = "extensions.shield-recipe-client.";
-const DEFAULT_PREFS = {
-  api_url: "https://normandy.cdn.mozilla.net/api/v1",
-  dev_mode: false,
-  enabled: true,
-  startup_delay_seconds: 300,
-  "logging.level": Log.Level.Warn,
-  user_id: "",
-  run_interval_seconds: 86400, // 24 hours
-};
-const PREF_DEV_MODE = "extensions.shield-recipe-client.dev_mode";
-const PREF_SELF_SUPPORT_ENABLED = "browser.selfsupport.enabled";
-const PREF_LOGGING_LEVEL = PREF_BRANCH + "logging.level";
-
-let shouldRun = true;
-let log = null;
-
-// These are exported for testing purposes only.
-this.EXPORTED_SYMBOLS = ["install", "startup", "shutdown", "uninstall"];
-
-this.install = function() {
-  // Self Repair only checks its pref on start, so if we disable it, wait until
-  // next startup to run, unless the dev_mode preference is set.
-  if (Preferences.get(PREF_SELF_SUPPORT_ENABLED, true)) {
-    Preferences.set(PREF_SELF_SUPPORT_ENABLED, false);
-    if (!Preferences.get(PREF_DEV_MODE, false)) {
-      shouldRun = false;
-    }
-  }
-};
+this.install = function() {};
 
 this.startup = async function() {
-  setDefaultPrefs();
-
-  // Setup logging and listen for changes to logging prefs
-  LogManager.configure(Services.prefs.getIntPref(PREF_LOGGING_LEVEL));
-  log = LogManager.getLogger("bootstrap");
-  Preferences.observe(PREF_LOGGING_LEVEL, LogManager.configure);
-  CleanupManager.addCleanupHandler(
-    () => Preferences.ignore(PREF_LOGGING_LEVEL, LogManager.configure));
-
-  if (!shouldRun) {
-    return;
-  }
-
-  // Initialize experiments first to avoid a race between initializing prefs
-  // and recipes rolling back pref changes when experiments end.
-  try {
-    await PreferenceExperiments.init();
-  } catch (err) {
-    log.error("Failed to initialize preference experiments:", err);
-  }
-
-  await RecipeRunner.init();
+  await ShieldRecipeClient.startup();
 };
 
 this.shutdown = function(data, reason) {
-  CleanupManager.cleanup();
-
-  if (reason === REASONS.ADDON_DISABLE || reason === REASONS.ADDON_UNINSTALL) {
-    Services.prefs.setBoolPref(PREF_SELF_SUPPORT_ENABLED, true);
-  }
+  ShieldRecipeClient.shutdown(reason);
 
+  // Unload add-on modules. We don't do this in ShieldRecipeClient so that
+  // modules are not unloaded accidentally during tests.
+  const log = LogManager.getLogger("bootstrap");
```

Comment:
We may or may not have called LogManager.configure() here, per https://bugzilla.mozilla.org/show_bug.cgi?id=1351465#c5 . I don't know if that should be a serious concern, but I wanted to flag this up.